File: projects/fasta_structs.py
from email.policy import default
import sys
from typing import Dict, List
from collections import defaultdict
from dataclasses import dataclass

@dataclass
class SeqRecord():

    name: str = ''
    sequence: str = ''

    def tally_kmer_observations(self, kmer_size: int) -> Dict[str,int]:
        ''' Performing a sliding window over each entry in the contig list, adding to each observation of k-mer. '''

        obs_dict = defaultdict(lambda: 0)

        for i in range(0, len(self.sequence) - kmer_size + 1):
            kmer = self.sequence[i:i+kmer_size]
            obs_dict[kmer] += 1

        return obs_dict

def main():

    # Unpack arguments
    input_file = sys.argv[1]
    kmer_size = parse_kmer_size(sys.argv[2])

    # Parse the sequence content
    record_list = parse_sequence_file(input_file)

    for seq_record in record_list:
        contig_observations = seq_record.tally_kmer_observations(kmer_size)

        print(seq_record.name)
        for kmer, freq in contig_observations.items():
            print(f"  K-mer '{kmer}': {freq}")

#region Functions

def parse_kmer_size(kmer_input: str, kmer_default: int=8) -> int:
    ''' Attempt to parse the user input to an int, returning a default value on failure. '''

    try:

        kmer_value = int(kmer_input)
        return kmer_value

    except:

        print(f"Cannot parse value '{kmer_input}' to an integer, using default value of {kmer_default}")
        return kmer_default

def parse_sequence_file(file_path: str) -> List[SeqRecord]:
    ''' Parse the content of the fasta file, returning a list of the contig sequences. Contig names are discarded. '''

    record_buffer = []
    with open(file_path, 'r') as fna_reader:
        
        for record_chunk in fna_reader.read().split('>')[1:]:

            contig_name, *seq_fragments = record_chunk.split('\n')
            record_buffer.append(
                SeqRecord(name=contig_name, sequence=''.join(seq_fragments).upper())
            )

    return record_buffer

def tally_kmer_observations(contig_list: List[str], kmer_size: int) -> Dict[str,int]:
    ''' Performing a sliding window over each entry in the contig list, adding to each observation of k-mer. '''

    obs_dict = defaultdict(lambda: 0)

    for contig_sequence in contig_list:

        for i in range(0, len(contig_sequence) - kmer_size + 1):

            kmer = contig_sequence[i:i+kmer_size]
            obs_dict[kmer] += 1

    return obs_dict

#endregion

if __name__ == '__main__':
    main()

File: projects/fasta_structs.rs
use std::io;
use std::env;
use std::fs::File;
use std::io::Read;
use std::iter::FromIterator;
use std::collections::HashMap;

fn main() {

    // This is not a complete rewrite, as parts o 

    // Capture the user inputs
    let file_path = env::args().nth(1).unwrap();
    let kmer_input = env::args().nth(2).unwrap();

    let kmer_size: usize = kmer_input.parse().expect("Cannot convert value to integer!");

    // Capture the file content
    let file_result = read_file_contents(&file_path);
    let record_vector: Option<Vec::<SeqRecord>> = if file_result.is_ok() {

        // Unpack the results into a vector of Strings (lines), then consume the line content into SeqRecord instances
        let file_content = file_result.unwrap();
        let file_lines: Vec<&str> = file_content.split("\n").collect();
        Some(parse_file_to_records(file_lines))
    } else {
        None
    };

    if record_vector.is_some() {

        for f in &record_vector.unwrap() {

            let kmer_observations = f.produce_kmer_tally(kmer_size);
            println!("{}", f.name);

            for (kmer, freq) in kmer_observations {
                println!("  K-mer '{}': {}", kmer, freq);
            }
        }
    }
}

//region File processing

fn read_file_contents(file_path: &str)-> io::Result<String> {
    let mut file_handle = File::open(&file_path)?;
    let mut file_content = String::new();
    file_handle.read_to_string(&mut file_content)?;
    Ok(file_content)
}

fn check_is_header(file_line: &str) -> bool {

    // Unwrap the value correctly if Some(), else add a poison pill.
    let ch = &file_line.chars().nth(0).unwrap_or('x');

    match ch {
        '>' => true,
        _ => false
    }
}

fn parse_file_to_records(file_lines: Vec<&str>) -> Vec<SeqRecord> {

    // Declare the results vector and the working struct for capturing data
    let mut record_vector: Vec<SeqRecord> = Vec::new();
    let mut current_record = SeqRecord::new("");

    for file_line in &file_lines {

        // Case: Encountered a header
        if check_is_header(file_line) {    

            // If the current fasta has been loaded, store it, then either way create a new empty record
            if current_record.is_loaded() {
                record_vector.push(current_record);
            }
            let sequence_name = &file_line[1..];
            current_record = SeqRecord::new(sequence_name);

        // Case: We are at a sequence line, or the end of the file
        } else {
            current_record.extend_sequence(file_line);
        }
    }

    // Close out the final entry
    if current_record.is_loaded() {
        record_vector.push(current_record);
    }

    record_vector
}


//endregion

// region SeqRecord

#[derive(Debug)]
struct SeqRecord {
    name: String,
    sequence: Vec<char>
}

impl SeqRecord {

    fn new(name: &str) -> SeqRecord {
        SeqRecord {
            name: name.to_string(),
            sequence: Vec::new(),
        }
    }

    fn extend_sequence(&mut self, new_sequence: &str) {

        // Push all sequences characters to uppercase, so there are't mismatches between 'a' and 'A'.
        let converted_nts = &new_sequence.to_uppercase();

        // Specifying the data type on the left hand saves me a turbofish on the right hand.
        let additional_nts: Vec<char> = converted_nts.chars().collect();

        self.sequence.extend(additional_nts);
    }

    fn is_loaded(&self) -> bool {
        if self.name != "".to_string() {
            true
        } else {
            false
        }
    }

    fn produce_kmer_tally(&self, window_size: usize) -> HashMap<String, i64> {

        // Declare a HashMap to hold the observations
        let mut kmer_observations: HashMap<String, i64> = HashMap::new();

        for kmer_array in self.sequence.windows(window_size) {

            let kmer = String::from_iter(kmer_array);
            *kmer_observations.entry(kmer).or_insert(0) += 1;
        }
        kmer_observations
    }
}

// endregion
